File: javascript/src/utils/ShellExecutor.ts
import { exec, execSync } from 'child_process';
import { promisify } from 'util';
import { Tool } from 'langchain/tools';
import { BaseTool } from '@langchain/core/tools';
import { CallbackManagerForToolRun } from '@langchain/core/callbacks/manager';
import { getLogger } from './LoggerSetup';

const execAsync = promisify(exec);
const logger = getLogger('ShellExecutor');

/**
 * Result interface for shell command execution
 */
export interface ShellCommandResult {
  success: boolean;
  returnCode: number | null;
  stdout: string;
  stderr: string;
  duration?: number;
  command: string;
  error?: string;
}

/**
 * Class for executing shell commands
 */
export class ShellExecutor {
  /**
   * Run a shell command asynchronously
   * 
   * @param command - The command to execute
   * @param timeout - Optional timeout in milliseconds
   * @param cwd - Optional working directory for the command
   * @param env - Optional environment variables
   * @returns Result of the command execution
   */
  static async runAsync(
    command: string,
    timeout?: number,
    cwd?: string,
    env?: NodeJS.ProcessEnv
  ): Promise<ShellCommandResult> {
    logger.info(`Executing command: ${command}`);
    const startTime = Date.now();
    
    try {
      // Create options object
      const options: any = {
        shell: true,
        maxBuffer: 10 * 1024 * 1024 // 10MB buffer
      };
      
      if (cwd) options.cwd = cwd;
      if (env) options.env = { ...process.env, ...env };
      if (timeout) options.timeout = timeout;
      
      // Execute the command
      const { stdout, stderr } = await execAsync(command, options);
      const duration = (Date.now() - startTime) / 1000;
      
      logger.debug(`Command completed in ${duration.toFixed(2)}s`);
      
      return {
        success: true,
        returnCode: 0,
        stdout: stdout,
        stderr: stderr,
        duration: duration,
        command: command
      };
    } catch (error: any) {
      const duration = (Date.now() - startTime) / 1000;
      logger.error(`Error executing command: ${error.message}`);
      
      // Determine error type
      let errorType = 'execution_error';
      if (error.signal === 'SIGTERM' && error.killed) {
        errorType = 'timeout';
      }
      
      return {
        success: false,
        returnCode: error.code || null,
        stdout: error.stdout || '',
        stderr: error.stderr || error.message,
        duration: duration,
        command: command,
        error: errorType
      };
    }
  }
  
  /**
   * Run a shell command synchronously
   * 
   * @param command - The command to execute
   * @param timeout - Optional timeout in milliseconds
   * @param cwd - Optional working directory for the command
   * @param env - Optional environment variables
   * @returns Result of the command execution
   */
  static run(
    command: string,
    timeout?: number,
    cwd?: string,
    env?: NodeJS.ProcessEnv
  ): ShellCommandResult {
    logger.info(`Executing command synchronously: ${command}`);
    
    try {
      // Create options object
      const options: any = {
        shell: true,
        maxBuffer: 10 * 1024 * 1024, // 10MB buffer
        encoding: 'utf8'
      };
      
      if (cwd) options.cwd = cwd;
      if (env) options.env = { ...process.env, ...env };
      if (timeout) options.timeout = timeout;
      
      // Execute the command
      const stdout = execSync(command, options).toString();
      
      logger.debug(`Command completed successfully`);
      
      return {
        success: true,
        returnCode: 0,
        stdout: stdout,
        stderr: '',
        command: command
      };
    } catch (error: any) {
      logger.error(`Error executing command synchronously: ${error.message}`);
      
      return {
        success: false,
        returnCode: error.status || null,
        stdout: error.stdout ? error.stdout.toString() : '',
        stderr: error.stderr ? error.stderr.toString() : error.message,
        command: command,
        error: 'execution_error'
      };
    }
  }
}

/**
 * LangChain tool for executing shell commands
 */
export class ShellExecutorTool extends BaseTool {
  name = 'shell_executor';
  description = 'Execute shell commands on the local system';
  
  constructor() {
    super();
  }
  
  /** @ignore */
  async _call(
    args: { command: string; timeout?: number; cwd?: string },
    runManager?: CallbackManagerForToolRun
  ): Promise<string> {
    const { command, timeout, cwd } = args;
    
    if (runManager) {
      await runManager.handleText(`Executing shell command: ${command}`);
    }
    
    const result = await ShellExecutor.runAsync(command, timeout, cwd);
    
    // Format the result for easier readability by the LLM
    const output = [
      `Command: ${command}`,
      `Success: ${result.success}`,
      `Return Code: ${result.returnCode}`,
      '',
      'STDOUT:',
      result.stdout,
      '',
      'STDERR:',
      result.stderr
    ].join('\n');
    
    return output;
  }
}

/**
 * Factory function to create a shell executor tool
 * @returns A configured ShellExecutorTool instance
 */
export function createShellExecutorTool(): ShellExecutorTool {
  return new ShellExecutorTool();
}

File: python/src/utils/shell_executor.py
import os
import subprocess
import asyncio
from typing import Dict, List, Any, Optional, Tuple, Union
from loguru import logger

from langchain_core.callbacks.manager import CallbackManagerForToolRun
from langchain_core.tools import BaseTool
from langchain.tools import tool


class ShellExecutor:
    """Utility for executing shell commands.
    
    This class provides methods for running shell commands synchronously
    and asynchronously, with support for timeouts and error handling.
    """
    
    @staticmethod
    async def run_async(
        command: str, 
        timeout: Optional[int] = None, 
        cwd: Optional[str] = None,
        env: Optional[Dict[str, str]] = None,
        shell: bool = True
    ) -> Dict[str, Any]:
        """Run a shell command asynchronously.
        
        Args:
            command: The command to execute
            timeout: Optional timeout in seconds
            cwd: Optional working directory for the command
            env: Optional environment variables
            shell: Whether to run through the shell
            
        Returns:
            Dictionary with command output details
        """
        logger.info(f"Executing command: {command}")
        start_time = asyncio.get_event_loop().time()
        
        try:
            # Create subprocess
            process = await asyncio.create_subprocess_shell(
                command,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE,
                cwd=cwd,
                env=env if env else os.environ,
                shell=shell
            )
            
            # Wait for the command to complete with timeout
            stdout, stderr = await asyncio.wait_for(
                process.communicate(), 
                timeout=timeout
            )
            
            end_time = asyncio.get_event_loop().time()
            duration = end_time - start_time
            
            stdout_str = stdout.decode('utf-8', errors='replace')
            stderr_str = stderr.decode('utf-8', errors='replace')
            
            logger.debug(f"Command completed in {duration:.2f}s with return code {process.returncode}")
            
            if process.returncode != 0:
                logger.warning(f"Command exited with non-zero code: {process.returncode}")
                logger.debug(f"stderr: {stderr_str}")
            
            return {
                "success": process.returncode == 0,
                "return_code": process.returncode,
                "stdout": stdout_str,
                "stderr": stderr_str,
                "duration": duration,
                "command": command
            }
            
        except asyncio.TimeoutError:
            logger.error(f"Command timed out after {timeout}s: {command}")
            return {
                "success": False,
                "return_code": None,
                "stdout": "",
                "stderr": "Command execution timed out",
                "duration": asyncio.get_event_loop().time() - start_time,
                "command": command,
                "error": "timeout"
            }
        except Exception as e:
            logger.error(f"Error executing command: {str(e)}")
            return {
                "success": False,
                "return_code": None,
                "stdout": "",
                "stderr": str(e),
                "duration": asyncio.get_event_loop().time() - start_time,
                "command": command,
                "error": "execution_error"
            }
    
    @staticmethod
    def run(
        command: str, 
        timeout: Optional[int] = None, 
        cwd: Optional[str] = None,
        env: Optional[Dict[str, str]] = None,
        shell: bool = True
    ) -> Dict[str, Any]:
        """Run a shell command synchronously.
        
        Args:
            command: The command to execute
            timeout: Optional timeout in seconds
            cwd: Optional working directory for the command
            env: Optional environment variables
            shell: Whether to run through the shell
            
        Returns:
            Dictionary with command output details
        """
        logger.info(f"Executing command synchronously: {command}")
        
        try:
            # Execute the command
            process = subprocess.run(
                command,
                shell=shell,
                cwd=cwd,
                env=env if env else os.environ,
                capture_output=True,
                text=True,
                timeout=timeout
            )
            
            logger.debug(f"Command completed with return code {process.returncode}")
            
            if process.returncode != 0:
                logger.warning(f"Command exited with non-zero code: {process.returncode}")
                logger.debug(f"stderr: {process.stderr}")
            
            return {
                "success": process.returncode == 0,
                "return_code": process.returncode,
                "stdout": process.stdout,
                "stderr": process.stderr,
                "command": command
            }
            
        except subprocess.TimeoutExpired:
            logger.error(f"Command timed out after {timeout}s: {command}")
            return {
                "success": False,
                "return_code": None,
                "stdout": "",
                "stderr": "Command execution timed out",
                "command": command,
                "error": "timeout"
            }
        except Exception as e:
            logger.error(f"Error executing command: {str(e)}")
            return {
                "success": False,
                "return_code": None,
                "stdout": "",
                "stderr": str(e),
                "command": command,
                "error": "execution_error"
            }


class ShellExecutorTool(BaseTool):
    """Tool for executing shell commands with LangChain."""
    
    name: str = "shell_executor"
    description: str = "Execute shell commands on the local system"
    
    def __init__(self):
        """Initialize the shell executor tool."""
        super().__init__()
        self.shell_executor = ShellExecutor()
        
    async def _arun(self, command: str, timeout: Optional[int] = None, 
                   cwd: Optional[str] = None,
                   run_manager: Optional[CallbackManagerForToolRun] = None) -> str:
        """Run the shell command asynchronously.
        
        Args:
            command: The command to execute
            timeout: Optional timeout in seconds
            cwd: Optional working directory
            run_manager: Optional callback manager
            
        Returns:
            Command execution results as a string
        """
        if run_manager:
            await run_manager.on_text("Executing shell command: " + command)
            
        result = await self.shell_executor.run_async(command, timeout, cwd)
        
        # Format the result for easier readability by the LLM
        output = (
            f"Command: {command}\n"
            f"Success: {result['success']}\n"
            f"Return Code: {result['return_code']}\n\n"
            f"STDOUT:\n{result['stdout']}\n\n"
            f"STDERR:\n{result['stderr']}\n"
        )
        
        return output


@tool
def execute_shell(command: str, timeout: Optional[int] = None, cwd: Optional[str] = None) -> str:
    """
    Execute a shell command on the local system.
    
    Args:
        command: The command to execute
        timeout: Optional timeout in seconds
        cwd: Optional working directory
        
    Returns:
        Command execution results
    """
    # Create a synchronous wrapper for the shell executor
    shell_executor = ShellExecutor()
    
    # Use the event loop to run the async function
    loop = asyncio.get_event_loop()
    result = loop.run_until_complete(shell_executor.run_async(command, timeout, cwd))
    
    # Format the result
    output = (
        f"Command: {command}\n"
        f"Success: {result['success']}\n"
        f"Return Code: {result['return_code']}\n\n"
        f"STDOUT:\n{result['stdout']}\n\n"
        f"STDERR:\n{result['stderr']}\n"
    )
    
    return output